lib/LowerAddrSpaceCastPass.cpp

```diff
@@ -350,20 +350,87 @@ Value *clspv::LowerAddrSpaceCastPass::visitPtrToIntInst(PtrToIntInst &I) {
   return ptrToInt;
 }
 
+static bool DependsOnPhiNode(const Value *V, const PHINode *PhiNode,
+                             DenseSet<const Value *> &Visited) {
+  if (!isa<Instruction>(V)) {
+    return false;
+  }
+
+  SmallVector<const Value *, 16> Stack;
+  Stack.push_back(V);
+
+  while (!Stack.empty()) {
+    const Value *Current = Stack.pop_back_val();
+    if (Visited.contains(Current)) {
+      continue;
+    }
+    Visited.insert(Current);
+
+    if (Current == PhiNode) {
+      return true;
+    }
+
+    if (const auto *Inst = dyn_cast<Instruction>(Current)) {
+      for (const auto *Op : Inst->operand_values()) {
+        if (!Visited.contains(Op)) {
+          Stack.push_back(Op);
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
 Value *clspv::LowerAddrSpaceCastPass::visitPHINode(llvm::PHINode &I) {
   IRBuilder<> B(&I);
-  // NOTE: We assume that the first incoming value does not depend on I.
-  auto *V1 = visit(I.getIncomingValue(0));
-  auto *B1 = I.getIncomingBlock(0);
 
-  // Register the replacement early in order to avoid recursive calls
-  // when an incoming value depends on this PHI node.
-  auto Phi = B.CreatePHI(V1->getType(), I.getNumIncomingValues());
+  unsigned N = I.getNumIncomingValues();
+
+  // Analyse the incoming values anche check whether the types agree.
+  // Ignore any incoming value that depends on the node itself, as that would
+  // lead to infinite recursion. Instead, we delay processing them until
+  // after we have registered the replacement.
+  SmallVector<Value *, 2> Replacements(N);
+
+  DenseSet<const llvm::Value *> VisitedForSelfDep;
```

Comment:
This variable is only used inside `DependsOnPhiNode`, I think it should be defined there, not here.

Reply:
This way we avoid re-allocating the storage for the DenseSet and it's cleared between calls. But I can move it into the function if you don't think it's worthwhile

Reply:
Keeping the code as readable as possible is better in my opinion.

Reply:
Ok, changed :)